rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,355 @@
+
+## Required Approvals
+
+- Engineering: @r0mant && @smallinsky
+- Product: @xinding33 || @klizhentas
+- Security: @reedloden || @jentfoo
+
+## What
+
+Extends [automated db user provisioning](0113-automatic-database-users.md) with
+permission management capabilities.
+
+## Why
+
+Managing database-level permissions is a natural extension of current Teleport
+RBAC capabilities. The described model should also integrate seamlessly with
+TAG, providing a future-proof solution.
+
+Database administrators will be able to use Teleport for both user and
+permission management.
+
+## Details
+
+### UX
+
+A set of global "database object import rules" resources are defined for the
+Teleport cluster. At certain points in time, the database schema is read and
+passed through the import rules. The individual import rules may apply custom
+labels. An database object resource will only be created if there is at least
+one import rule that matched it.
+
+Only object attributes (standard ones like `protocol` or custom from the
+`attributes` field), which are sourced from the object spec, are subject to
+matching against the import rules. Any labels present on an object, either from
+an import rule or another source, are not subject to be matched by the import
+rule. The permissions matching is different in this regard.
+
+Example: the `sales-prod` import rule, which imports all tables from `sales`
+Postgres database in prod:
+
+```yaml
+kind: db_object_import_rule
+metadata:
+  name: sales-prod
+spec:
+  db_labels:
+    env: prod
+  mappings:
+    - object_match:
+        - database: sales
+          object_kind: table
+          protocol: postgres
+      add_labels:
+        env: prod
+        product: sales
+  priority: 10
+version: v1
+```
+
+Example database object, with applied labels:
+
+```yaml
+kind: db_object
+metadata:
+  labels:
+    env: prod
+    product: sales
+  name: sales_main
+spec:
+  attributes:
+    attr1: custom attr1 value
+    attr2: custom attr2 value
+  database: sales
+  name: sales_main
+  object_kind: table
+  protocol: postgres
+  schema: public
+  service_name: sales-prod-123
+version: v1
+```
+
+The permissions to particular objects are defined in a role using the new field
+`db_permissions`. The permission specifies the object kind (e.g. `table`),
+permission to grant/revoke (`SELECT`) as well as labels to be matched.
+
+The matching is performed against database object:
+
+- attributes: standard and custom, sourced from the object spec, provided by
+  database-specific schema import code
+- resource labels: aside from standard labels, these can be manipulated by the
+  import rules.
+
+If both attributes and labels provide the same non-empty key, the attributes
+take preference.
+
+Example role `db-dev`, which grants `SELECT` permission to some objects, and
+revokes `UPDATE` from all others.
+
+```yaml
+kind: role
+metadata:
+  name: db-dev
+spec:
+  allow:
+    db_permissions:
+      - match:
+          product: sales
```

Comment:
It's unclear that this is matching label `product: sales`. It's also on the same level as matching `object_kind`. What if there is a label `object_kind` that conflicts?

We probably need an extra level:

```
match:
  labels:
     product: sales
  object_kind: table
  db_type: postgres
```
or something similar

Reply:
In current design attributes and labels are matched together; if the same key is present for both, then the attributes are taking precedence. But I can see this is causing confusion, so its probably best that I change that. This will also remove some ambiguity, while the verbosity increase should be minimal.

Reply:
I've removed the attributes from the `db_permissions` in roles; now this is purely based on the object labels.